Proposed LGC interface for shader compilation and linking

trenouf opened the pull request:
The top two commits define an LGC interface for shader compilation generating an unlinked ELF, and ELF linking.

For now this is only for discussion.

amdrexu commented:
I might need an interface that could extract GPU codes from a shader module ELF (just shader module, not pipeline) compiled by back-end and inject it to a patcher (independent of LLPC/LGC) that will insert additional "harmless" data dump instructions (buffer stores) and reconstruct this shader module ELF. Extra processing might be the modifications of jump label symbols (BB_X and its offsets), certain PAL ABI metadata, and shader registers (such as VGPR and SGPR usage).

This could be done within LGC as a patch pass. Or be done when front-end get the compilation results and modify it. I guess the latter is not elegant and need builder to provide additional interfaces.

nhaehnle commented on lgc/interface/lgc/ElfLinker.h:
```diff
@@ -0,0 +1,84 @@
+  // @param glueIndex : Index into the array that was returned by getGlueInfo()
+  // @return : The blob. A zero-length blob indicates that a recoverable error occurred, and link() will also return
+  //           and empty ELF blob.
+  virtual llvm::StringRef compileGlue(unsigned glueIndex) = 0;
```
StringRef isn't really the right type for this, is it?

trenouf replied:
It seems to be the done thing. It is common to pass around a MemoryBufferRef, which is a StringRef with an attached name for error reporting etc. So the use of StringRef to represent a read-only binary blob whose memory you don't own does seem right.

nhaehnle replied:
Ah I see, you're right.

nhaehnle commented on lgc/interface/lgc/ElfLinker.h:
```diff
@@ -0,0 +1,84 @@
+  // is a StringRef that the client can hash for its cache lookup. If it gets a cache hit, it should provide the
+  // found blob to ElfLinker::addGlue. If it does not get a cache hit, the client can call ElfLinker::compileGlue to
+  // retrieve the compiled glue code to store in the cache.
+  virtual llvm::ArrayRef<llvm::StringRef> getGlueInfo() = 0;
```
Should this just be a hash then instead of a StringRef?

trenouf replied:
I did wonder about that, but that ties the front-end (and/or driver) into a particular size of hash that might disagree with other stuff it does. So I thought it would be better to let the front-end do the hashing in the same way it does other hashing.

nhaehnle replied:
Fair enough.

nhaehnle commented on lgc/interface/lgc/Pipeline.h:
```diff
@@ -581,17 +582,32 @@ class Pipeline {
+  // 2. Return false for recoverable failure because irLink asked for an "unlinked" half-pipeline and there is some
+  //    reason why the module cannot be compiled in that way. The client typically then does a whole-pipeline
+  //    compilation instead. (This cannot happen if the call to irLink did not ask for an unlinked half-pipeline.)
+  // 3. Other failures cause exit by report_fatal_error, which the client can catch by setting a diagnostic handler
+  //    with LLVMContext::setDiagnosticHandler.
```
Other failures shouldn't really happen if the input is valid. Perhaps out-of-memory conditions, but can those really be caught with setDiagnosticHandler?

If false means "unlinked" didn't work, perhaps a Result enum would be clearer here.

trenouf replied:
Yes, other failures tend to be because input is invalid, which llpc does already catch. I guess when they're most useful (and not caught, but just reported to stderr) is in a standalone tool.

trenouf replied:
I don't want to do a Result enum as it seems over-the-top for the two conditions we can return here ("success" or "you need to do full pipeline compile").

LLVM generally does not provide a recoverable error mechanism from running a pass, so we can just assume that any error that wasn't supposed to happen (e.g. a pass thinks that something is invalid) will cause a report_fatal_error, or an assert, or setting your computer on fire.

nhaehnle replied:
Can we just reword the comment then? "Other failures" sounds too much like the client should *expect* them and *must* handle them, when really, they're internal compiler errors that should never happen.

So I wouldn't even list them as a 3rd item here, just maybe a note that internal compiler errors are reported as assertions and/or via llvm's report_fatal_error.

nhaehnle commented on lgc/state/Compiler.cpp:
```diff
@@ -128,17 +128,31 @@ Module *PipelineState::irLink(ArrayRef<std::pair<Module *, ShaderStage>> modules
+// 2. Return false for recoverable failure because irLink asked for an "unlinked" half-pipeline and there is some
+//    reason why the module cannot be compiled in that way. The client typically then does a whole-pipeline
+//    compilation instead.
+// 3. Other failures cause exit by report_fatal_error, which the client can catch by setting a diagnostic handler
+//    with LLVMContext::setDiagnosticHandler.
```
Similar concern here about "other failures".

trenouf commented:
V2: Added early check for shader compilability. Added error message
        getting. link method now takes raw_pwrite_stream. Changed
        Pipeline::getElfLinker to Pipeline::createElfLinker, with caller
        responsible for freeing.
    V3: Addressed review comment.

trenouf commented:
V4: Fixed clang-format.

trenouf commented:
Ping for approval, as I need to merge this before #673.

trenouf commented:
Rebased

trenouf commented:
V5: Fixed clang-format again.
    V6: Fixed lgc.cpp compile failure.

amdvlk-admin commented:
### Test summary for commit 8299e876538e6e01bc892d542cc2b61cc71b7e3b
<details open><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master b601c88aeca7a7b08becb3d32709de383c8ee428</li>
<li>METROHASH: amd-master 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev ddcaa1009ecde219613042a5a53c6d3752bf4b9d</li>
<li>SPVGEN: dev d4817ab957f79762da83b7fc1d3f2816dbdc13fd</li>
<li>XGL: dev 2ecd2e14e2ae4f2712df4a2220e2dd36265c66d1</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c57487b76215ee74b1038039f47b210a28cad65e</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 960/960 (100.0%)</li>
<li>Failed: 0/960 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 103143/103143 (100.0%)</li>
<li>Failed: 0/103143 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Navi10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 103199/103199 (100.0%)</li>
<li>Failed: 0/103199 (0.0%)</li>
</ul></details>


</ul></details>
